cpp/src/arrow/types/struct.cc

```diff
@@ -17,4 +17,94 @@
 
 #include "arrow/types/struct.h"
 
-namespace arrow {}  // namespace arrow
+#include <sstream>
+
+namespace arrow {
+
+bool StructArray::EqualsExact(const StructArray& other) const {
+  if (this == &other) { return true; }
+  if (null_count_ != other.null_count_) { return false; }
+
+  bool equal_null_bitmap = true;
+  if (null_count_ > 0) {
+    equal_null_bitmap =
+        null_bitmap_->Equals(*other.null_bitmap_, util::bytes_for_bits(length_));
+  }
+
+  if (!equal_null_bitmap) { return false; }
+  bool values_equal = false;
+  for (size_t i = 0; i < field_arrays_.size(); ++i) {
+    values_equal = field_arrays_.at(i).get()->Equals(other.fields().at(i));
+  }
+
+  return values_equal;
+}
+
+bool StructArray::Equals(const std::shared_ptr<Array>& arr) const {
+  if (this == arr.get()) { return true; }
+  if (this->type_enum() != arr->type_enum()) { return false; }
+  return EqualsExact(*static_cast<const StructArray*>(arr.get()));
+}
+
+Status StructArray::Validate() const {
+  if (length_ < 0) { return Status::Invalid("Length was negative"); }
+
+  if (null_count() > length_) {
+    return Status::Invalid("Null count exceeds the length of this struct");
+  }
+
+  if (fields().size() <= 0 && length_ > 0) {
+    return Status::Invalid("Fields are not existed, but length is not zero");
+  }
+
+  int32_t struct_length = length();
```

Comment:
I don't think you need to this check.  the loop inside of it works either way, doesn't it?  and I think it would miss the case where stuct length is zero and fields have positive lengths.